sat/recipe.py

```diff
@@ -0,0 +1,178 @@
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+"""
+Defines a class for accessing the HPC software recipe manifest.
+
+The manifest defines the versions of the products that are included in a given
+version of the HPC software recipe.
+"""
+from contextlib import contextmanager
+from distutils.version import LooseVersion
+from functools import total_ordering
+import logging
+import os
+import re
+from tempfile import TemporaryDirectory
+from urllib.parse import ParseResult, urlunparse
+
+import yaml
+
+from sat.apiclient.vcs import VCSError, VCSRepo
+from sat.cached_property import cached_property
+from sat.config import get_config_value
+
+HPC_SOFTWARE_RECIPE_REPO_ORG = 'cray'
+HPC_SOFTWARE_RECIPE_REPO_NAME = 'hpc-software-recipe'
+HPC_SOFTWARE_RECIPE_REPO_PATH = f'/vcs/{HPC_SOFTWARE_RECIPE_REPO_ORG}/{HPC_SOFTWARE_RECIPE_REPO_NAME}.git'
+# Be generous with what is allowed after cray/hpc-software-recipe/
+HPC_SOFTWARE_RECIPE_RELEASE_REGEX = rf'refs/heads/(cray/{HPC_SOFTWARE_RECIPE_REPO_NAME}/(.+))'
+HPC_SOFTWARE_RECIPE_VARS_FILE = 'product_vars.yaml'
+
+LOGGER = logging.getLogger(__name__)
+
+
+class HPCSoftwareRecipeError(Exception):
+    """An error occurred querying the software recipes on the system."""
+
+
+@total_ordering
+class HPCSoftwareRecipe:
+
+    def __init__(self, version, vcs_repo, vcs_branch):
+        """Create a new HPCSoftwareRecipe object.
+
+        Args:
+            version (str): the recipe version string
+            vcs_repo (VCSRepo): the VCS repository containing software recipe information
+            vcs_branch: the full branch name of the `vcs_repo` that defines this recipe version
+        """
+        # Allows for comparison of two software recipes. Original string can be
+        # retrieved by calling `str` on it.
+        self.version = LooseVersion(version)
+        self.vcs_repo = vcs_repo
+        self.vcs_branch = vcs_branch
+
+    def __eq__(self, other):
+        return self.version == other.version
+
+    def __lt__(self, other):
+        return self.version < other.version
+
+    @cached_property
+    def all_vars(self):
+        """dict: the variables defined by the products"""
+        with self.clone_content() as clone_dir:
```

Comment:
This would likely require more instrumentation around the Gitea API, but we could potentially use the [`/repos/{owner}/{repo}/contents/{filepath}`](https://try.gitea.io/api/swagger#/repository/repoGetContents) endpoint from Gitea to retrieve just the contents of the recipe file, instead of the whole repository. Cloning the whole repo could potentially slow things down depending on how large it gets or how long the history becomes, though depending on how small it stays this could also be a premature optimization.

Reply:
Agree that this would be a nice future optimization. I do have some future hopes for which clone functionality would be useful. Specifically, it would be nice to have some automation around helping the admin create and merge the `integration` branches of these repos, perhaps as another mode of `sat bootprep`.